src/generic.jl

```diff
@@ -109,17 +109,48 @@ function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractMatrix)
     size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r."))
     @inbounds for j = 1:n
         aj = view(a, :, j)
+        for i = 1:(j - 1)
+            r[i, j] = r[j, i]   # leveraging the symmetry of SemiMetric
+        end
+        r[j, j] = 0
         for i = (j + 1):n
             r[i, j] = metric(view(a, :, i), aj)
         end
-        r[j, j] = 0
+    end
+    r
+end
+
+function _pairwise!(r::AbstractMatrix, metric::PreMetric,
+                    a::AbstractVector, b::AbstractVector=a)
```

Comment:
With the default value the 3-arg method is defined twice:
```suggestion
function _pairwise!(r::AbstractMatrix, metric::PreMetric,
                    a::AbstractVector, b::AbstractVector)
```

Reply:
Yes because it goes then to the specialized method `_pairwise(r, metric, a)` which takes advantage of the symmetry

Reply:
What I mean is that if you keep the default value `b=a` you define `_pairwise(r, metric, a)` twice.

Reply:
Oh right! Sorry. Well actually it's also true for the AbstractMatrix case... I will correct both

Reply:
Ok figured out why it's like this. The one argument method is only for `SemiMetric`... I switched it back